project1/server/agent.cpp

```diff
@@ -4,231 +4,416 @@
+
+        // Send the response.
+        Response response;
+        response.mutable_list()->clear_filenames();
+        for (const auto &kFilename : kDirectoryContents) {
+            response.mutable_list()->add_filenames(kFilename);
+        }
+        return SendResponse(response) ? ClientState::ACTIVE : ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::HandleRequest(
+            const ftp_messages::ChangeDirRequest &request) {
+        std::cout << "Performing a CD request." << std::endl;
+
+        if (request.go_up()) {
+            // Go up a directory.
+            if (!file_handler_->UpDir()) {
+                std::cerr << "Failed to go up a directory." << std::endl;
+                return ClientState::ERROR;
+            }
+        } else {
+            // Go down a directory.
+            if (!file_handler_->ChangeDir(request.dir_name())) {
+                std::cerr << "Failed to go down a directory." << std::endl;
+                return ClientState::ERROR;
+            }
+        }
+
+        return SendResponse(kEmptyResponse) ? ClientState::ACTIVE
+                                            : ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::HandleRequest(
+            const ftp_messages::MakeDirRequest &request) {
+        std::cout << "Performing a MKDIR request." << std::endl;
+
+        // Create the directory.
+        if (!file_handler_->MakeDir(request.dir_name())) {
+            std::cerr << "Failed to create a directory." << std::endl;
+            return ClientState::ERROR;
+        }
+
+        return SendResponse(kEmptyResponse) ? ClientState::ACTIVE
+                                            : ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::HandleRequest(
+            const ftp_messages::PwdRequest &request) {
+        std::cout << "Performing a PWD request." << std::endl;
+
+        // Get the directory.
+        const auto &kCurrentDir = file_handler_->GetCurrentDir();
+
+        // Send the response.
+        Response response;
+        response.mutable_pwd()->set_dir_name(kCurrentDir);
+        return SendResponse(response) ? ClientState::ACTIVE : ClientState::ERROR;
+    }
+
+    Agent::ClientState Agent::HandleRequest(
+            const ftp_messages::TerminateRequest &request) {
+        std::cout<< "Terminating an active GET or PUT command." << std::endl;
+
+        // Retrieve the command id.
+        std::string cmd_str = request.command_id();
+
+        // Case to uint16_t
+        int cmd_int(std::stoi(cmd_str));
+        uint16_t cmd_id(0);
+
+        if (cmd_int <= static_cast<int>(UINT16_MAX) && cmd_int >=0) {
+            cmd_id = static_cast<uint16_t>(cmd_int);
+        }
+        else {
+            std::cout << "Casting Error. " << std::endl;
+        }
+
+        // remove command from active command list.
+        active_commands_->Delete(cmd_id);
+
+        return ClientState::ACTIVE;
+
+    }
+
+    Agent::ClientState Agent::HandleRequest(
```

Comment:
The current client code doesn't ever send a `QuitRequest`, though it probably should.

Either this is unnecessary, or the client should be telling the server it is disconnecting. I think the later makes more sense.

Reply:
I agree. The client code should really send this request when the user types "quit". I think that behavior is actually a requirement for the project (or it was at least a requirement for project 1).